include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -0,0 +1,1018 @@
+    SizeType tile_row = (i - i_begin) * nb;
+    whenAllLift(tile_row, index.readwrite_sender(tile_idx)) |
+        initIndexTile(Policy<Backend::MC>(thread_priority::normal)) | start_detached();
+  }
+}
+
+// Copies and normalizes a row of the `tile` into the column vector tile `col`
+//
+template <class T>
+void copyTileRowAndNormalize(bool top_tile, T rho, const matrix::Tile<const T, Device::CPU>& tile,
+                             const matrix::Tile<T, Device::CPU>& col) {
+  // Copy the bottom row of the top tile or the top row of the bottom tile
+  SizeType row = (top_tile) ? col.size().rows() - 1 : 0;
+
+  // Negate Q1's last row if rho < 0
+  //
+  // lapack 3.10.0, dlaed2.f, line 280 and 281
+  int sign = (top_tile && rho < 0) ? -1 : 1;
+
+  for (SizeType i = 0; i < tile.size().cols(); ++i) {
+    col(TileElementIndex(i, 0)) = sign * tile(TileElementIndex(row, i)) / std::sqrt(2);
+  }
+}
+
+DLAF_MAKE_CALLABLE_OBJECT(copyTileRowAndNormalize);
+DLAF_MAKE_SENDER_ALGORITHM_OVERLOADS(copyTileRowAndNormalize, copyTileRowAndNormalize_o)
+
+// The bottom row of Q1 and the top row of Q2. The bottom row of Q1 is negated if `rho < 0`.
+//
+// Note that the norm of `z` is sqrt(2) because it is a concatination of two normalized vectors. Hence to
+// normalize `z` we have to divide by sqrt(2). That is handled in `copyTileRowAndNormalize()`
+template <class T>
+void assembleZVec(SizeType i_begin, SizeType i_split, SizeType i_end, pika::shared_future<T> rho_fut,
+                  Matrix<const T, Device::CPU>& mat_ev, Matrix<T, Device::CPU>& z) {
+  using dlaf::internal::Policy;
+  using dlaf::internal::whenAllLift;
+  using pika::execution::experimental::start_detached;
+  using pika::threads::thread_priority;
+
+  // Iterate over tiles of Q1 and Q2 around the split row `i_middle`.
+  for (SizeType i = i_begin; i <= i_end; ++i) {
+    // True if tile is in Q1
+    bool top_tile = i <= i_split;
+    // Move to the row below `i_middle` for `Q2`
+    SizeType mat_ev_row = i_split + ((top_tile) ? 0 : 1);
+    GlobalTileIndex mat_ev_idx(mat_ev_row, i);
+    // Take the last row of a `Q1` tile or the first row of a `Q2` tile
+    GlobalTileIndex z_idx(i, 0);
+    // Copy the row into the column vector `z`
+    whenAllLift(top_tile, rho_fut, mat_ev.read_sender(mat_ev_idx), z.readwrite_sender(z_idx)) |
+        copyTileRowAndNormalize(Policy<Backend::MC>(thread_priority::normal)) | start_detached();
+  }
+}
+
+// Multiply by factor 2 to account for the normalization of `z` vector and make sure rho > 0 f
+//
+template <class T>
+pika::future<T> scaleRho(pika::shared_future<T> rho_fut) {
+  // Note: `keep_future()` is needed here even though `T` is a scalar type (not a `Tile<>`) otherwise there
+  // is a compile-time error. A fix will be available in pika@0.6.0 : https://github.com/pika-org/pika/pull/282
+  namespace ex = pika::execution::experimental;
+  namespace di = dlaf::internal;
+  return ex::keep_future(std::move(rho_fut)) |
```

Comment:
This should work without `keep_future`. Did you find that you needed it to make things compile? If yes, that may indicate a bug. A scalar `T` should definitely be copiable from a `const&`.

Reply:
Locally it does compile for me without `keep_future()` but that is because I built `pika@0.5.0` with a patch containing a part of https://github.com/pika-org/pika/pull/282 (merged for `pika@0.6.0`). The error I reported to you a few days ago on slack was this : https://pastebin.com/uaGzmsuw , then you suggested to check if https://github.com/pika-org/pika/pull/282 will help and it did.

Reply:
Ah, sorry... it's even in the comment right above :facepalm: I remember now what this was about. Opened https://github.com/eth-cscs/DLA-Future/issues/590 for my own sake.